File: shared/python/src/dronefleet_shared/models/mission.py
from datetime import datetime

from pydantic import BaseModel

from .protocol import WaypointType
from .telemetry import Position


class Waypoint(BaseModel):
    type: WaypointType
    position: Position
    related_order_id: str | None = None
    related_warehouse_id: str | None = None


class Mission(BaseModel):
    id: str
    drone_id: str
    order_ids: list[str] = []
    route: list[Waypoint] = []
    status: str = "ACTIVE"  # ACTIVE, COMPLETED, FAILED
    start_time: datetime | None = None
    end_time: datetime | None = None
    estimated_battery_consumption: float | None = None
    estimated_duration_minutes: float | None = None

File: shared/java/src/main/java/com/dronefleet/shared/models/Mission.java
package com.dronefleet.shared.models;

import java.time.Instant;
import java.util.List;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Mission {
    private String id;
    private String droneId;
    private List<String> orderIds;
    private List<Waypoint> route;
    private String status; // ACTIVE, COMPLETED, FAILED
    private Instant startTime;
    private Instant endTime;
    private Double estimatedBatteryConsumption;
    private Double estimatedDurationMinutes;

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Waypoint {
        private WaypointType type;
        private Position position;
        private String relatedOrderId;
        private String relatedWarehouseId;
    }

    public enum WaypointType {
        DEPOT_START,
        WAREHOUSE_PICKUP,
        HOSPITAL_DELIVERY,
        DEPOT_RETURN
    }
}
